### PR TITLE
Use minimal proxies in ICA router

### PR DESCRIPTION
**Pull request by yorhodes**
### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1413

### Backward compatibility

Not compatible with existing ICA router

### Testing

Unit Tests

### CODE REVIEW

File: solidity/contracts/libs/MinimalProxy.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+// https://eips.ethereum.org/EIPS/eip-1167
+
+library MinimalProxy {
+    function bytecode(address implementation)
+        internal
+        pure
+        returns (bytes memory)
+    {
+        bytes10 creation = 0x3d602d80600a3d3981f3;
```
Comment (nambrot) [inline solidity/contracts/libs/MinimalProxy.sol:12]: What are these magical constants?

Comment (yorhodes) [reply solidity/contracts/libs/MinimalProxy.sol:12]: linked in the comment

Comment (asaj) [inline solidity/contracts/libs/MinimalProxy.sol:12]: nit: `_` prefixes for consistency

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 2]
```diff
@@ -60,17 +61,20 @@ contract InterchainAccountRouter is Router, IInterchainAccountRouter {
         view
         returns (address)
     {
-        return _getInterchainAccount(_salt(_origin, _sender));
+        bytes memory bytecode = MinimalProxy.bytecode(implementation);
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:64]: Why not move this line to _getInterchainAccount?

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 3]
```diff
@@ -84,11 +88,12 @@ contract InterchainAccountRouter is Router, IInterchainAccountRouter {
         return bytes32(abi.encodePacked(_origin, _sender));
     }
 
-    function _getInterchainAccount(bytes32 salt)
+    function _getInterchainAccount(bytes32 salt, bytes memory bytecode)
         internal
         view
         returns (address)
     {
+        bytes32 bytecodeHash = keccak256(bytecode);
```
Comment (asaj) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:96]: Should we write this to storage? The implementation address/bytecode can't change, can it?

**Review by asaj — APPROVED** [review]

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 2 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:64]: because it's used in the `Create2.deploy`

**Pull request merged**